File: lib/data/models/job_application.dart
import 'package:objectbox/objectbox.dart';

/// Application status progression.
/// Stored as int in ObjectBox; use [JobStatus.fromCode] / [.code] to convert.
enum JobStatus {
  applied(0),
  onlineTest(1),
  interview(2),
  offer(3),
  rejected(4),
  unknown(99);

  const JobStatus(this.code);
  final int code;

  static JobStatus fromCode(int code) =>
      JobStatus.values.firstWhere((s) => s.code == code, orElse: () => unknown);
}

@Entity()
class JobApplication {
  @Id()
  int id = 0;

  String company = '';

  @Property(type: PropertyType.date)
  DateTime appliedAt = DateTime.now();

  /// Stored as [JobStatus.code].
  int statusCode = JobStatus.applied.code;

  /// FK to EmailMessage.id that triggered this record.
  int emailId = 0;

  /// True when the user has manually overridden AI-extracted data.
  bool manuallyEdited = false;

  /// JSON-encoded list of {status, timestamp} objects for the timeline.
  /// Example: [{"status":0,"ts":"2026-06-15T09:00:00Z"}]
  String timelineJson = '[]';

  JobStatus get status => JobStatus.fromCode(statusCode);
  set status(JobStatus s) => statusCode = s.code;
}

File: backend/app/models/job_application.py
"""A job application extracted from email (Phase 3). Status is stored as an int
code (parity with the Flutter JobStatus enum); the timeline is JSON text."""

from __future__ import annotations

from datetime import datetime
from enum import IntEnum

from sqlalchemy import Boolean, DateTime, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from ..db import Base


class JobStatus(IntEnum):
    APPLIED = 0
    ONLINE_TEST = 1
    INTERVIEW = 2
    OFFER = 3
    REJECTED = 4
    UNKNOWN = 99

    @classmethod
    def from_code(cls, code: int) -> "JobStatus":
        try:
            return cls(code)
        except ValueError:
            return cls.UNKNOWN


class JobApplication(Base):
    __tablename__ = "job_application"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    company: Mapped[str] = mapped_column(String, default="")
    applied_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    status_code: Mapped[int] = mapped_column(Integer, default=JobStatus.APPLIED.value)

    # FK to EmailMessage.id that triggered this record (0 = manual entry).
    email_id: Mapped[int] = mapped_column(Integer, default=0, index=True)

    # True once the user overrides AI-extracted data — extraction won't clobber it.
    manually_edited: Mapped[bool] = mapped_column(Boolean, default=False)

    # JSON list of {"status": int, "ts": ISO8601} for the status timeline.
    timeline_json: Mapped[str] = mapped_column(Text, default="[]")
